### PR TITLE
UpdateModelValueCommand Enhancements

### PR DESCRIPTION
**Pull request by Benglin**
Background
-----
There is [a user request](http://adsk-oss.myjetbrains.com/youtrack/issue/MAGN-5738) to support modifying multiple nodes (in selection set) in one go. There is currently no good way to do this through `UpdateModelValueCommand` because it is meant to work only for a given `ModelBase` object. This pull request enhances the `UpdateModelValueCommand` to support multiple node updates in one go.

Changes
-----
- A new overloaded constructor is added to `UpdateModelValueCommand` to take `IEnumerable<Guid> modelGuids` instead of single `Guid modelGuid` (which is still used by other existing codes).

- `UpdateModelValueCommand.DeserializeCore` method is updated to deal with both legacy `XmlElement` and the new format (which stores multiple `Guid` as child `XmlElement`).

- In `WorkspaceModel.UpdateModelValue` method, we no longer throw `InvalidOperationException` if a subset of the retrieved `ModelBase` object cannot handle the `UpdateValueCore` call. For example, selecting a `NodeModel` and `NoteModel` then perform a set-lacing command, `NoteModel` will not be able to handle the `UpdateValueCore` method call, which is a valid use case.

Unit tests
-----
The `UpdateModelValueCommand` is being tested by nearly all recorded tests. In addition to those, a new test case `TestUpdateModelValueCommand1` is added to make sure it works with multiple `System.Guid` at once.

All **155 recorded test cases** are passing/failing as expected (as compared to `master`).

Reviewers
-----
Hi @aparajit-pratap, please take a look. Thanks!

### CODE REVIEW

File: src/DynamoCoreWpf/Views/CodeBlocks/CodeBlockEditor.xaml.cs [hunk 1]
```diff
@@ -437,7 +437,7 @@ private void CommitChanges(UndoRedoRecorder recorder)
             {
```
Comment (Benglin) [inline src/DynamoCoreWpf/Views/CodeBlocks/CodeBlockEditor.xaml.cs]: Unwanted comments removal.

File: src/DynamoCore/Models/WorkspaceModel.cs [hunk 2]
```diff
@@ -811,44 +811,22 @@ internal void SendModelEvent(Guid modelGuid, string eventName)
...
-                    "Model GUID: {1}\n" +
-                    "Property name: {2}\n" +
-                    "Property value: {3}",
-                    type, modelGuid, propertyName, value);
-
-                // All 'UpdateValue' calls must be handled by one of the 
-                // ModelBase derived classes that the 'UpdateModelValue'
-                // is intended for.
-                throw new InvalidOperationException(message);
-            }
```
Comment (Benglin) [inline src/DynamoCore/Models/WorkspaceModel.cs]: We no longer throw an exception if a subset of the retrieved `ModelBase` object cannot handle the `UpdateValueCore` call. For example, selecting a `NodeModel` and `NoteModel` then perform a set-lacing command, `NoteModel` will not be able to handle the `UpdateValueCore` method call.

File: src/DynamoCore/Models/WorkspaceModel.cs [hunk 3]
```diff
@@ -15,7 +15,7 @@
 using ProtoCore.AST;
 using String = System.String;
 using Utils = Dynamo.Nodes.Utilities;
-using NodeModificationUndoHelper = Dynamo.Core.UndoRedoRecorder.NodeModificationUndoHelper;
+using ModelModificationUndoHelper = Dynamo.Core.UndoRedoRecorder.ModelModificationUndoHelper;
```
Comment (Benglin) [inline src/DynamoCore/Models/WorkspaceModel.cs]: Just a name change (to use `Model`).

File: src/DynamoCore/Core/UndoRedoRecorder.cs [hunk 4]
```diff
@@ -448,21 +448,27 @@ public void Dispose()
...
         {
-            private readonly NodeModel node;
+            private readonly List<ModelBase> models;
             private readonly UndoRedoRecorder recorder;
             private readonly Dictionary<Guid, XmlElement> existingConnectors;
             private readonly Dictionary<Guid, ConnectorModel> remainingConnectors;
 
-            public NodeModificationUndoHelper(UndoRedoRecorder recorder, NodeModel node)
+            public ModelModificationUndoHelper(UndoRedoRecorder recorder, ModelBase model)
+                : this(recorder, new [] { model })
```
Comment (aparajit-pratap) [inline src/DynamoCore/Core/UndoRedoRecorder.cs]: Just curious why has `NodeModel` been replaced with `ModelBase`. Is it to support `NoteModel` updates? If so, why was it not handled earlier?

Comment (Benglin) [inline src/DynamoCore/Core/UndoRedoRecorder.cs]: Good question there, @aparajit-pratap! The `NodeModificationUndoHelper` was initially designed to deal with anything that has connectors (which turns out to be just `NodeModel`). The idea of that class was to help record node modification which can eventually result in dropping connectors. Now that we want to apply model update to a selection set, which can potentially include `NoteModel`, this class then gets extended to work with the base class `ModelBase`. Does that make sense?

File: src/DynamoCore/Models/RecordableCommands.cs [hunk 5]
```diff
@@ -931,17 +939,35 @@ public UpdateModelValueCommand(Guid modelGuid, string name, string value)
...
+                else
+                {
+                    // Parsing a new type of 'UpdateModelValueCommand' that works 
+                    // for multiple 'NodeModel' whose Guid values are each stored 
+                    // as a 'ModelGuid' child element under the main element.
+                    // 
+                    var modelGuids = (from XmlNode xmlNode in element.ChildNodes
+                                      where xmlNode.Name.Equals("ModelGuid")
+                                      select Guid.Parse(xmlNode.InnerText)).ToList();
+
```
Comment (aparajit-pratap) [inline src/DynamoCore/Models/RecordableCommands.cs]: fancy Linq!

File: src/Libraries/Samples/SampleLibraryUI/SampleLibraryUI.csproj [hunk 6]
```diff
@@ -4,7 +4,7 @@
   <PropertyGroup>
     <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
     <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
-    <ProjectGuid>{227B776D-8866-49E8-8C93-809CCF86AF12}</ProjectGuid>
+    <ProjectGuid>{0A4B4EEA-8FAB-4AC8-90D4-27DBC5B0CF2A}</ProjectGuid>
     <OutputType>Library</OutputType>
```
Comment (aparajit-pratap) [inline src/Libraries/Samples/SampleLibraryUI/SampleLibraryUI.csproj]: I think you may need to revert this file.

**Comment by aparajit-pratap** [comment]
Apart from that one file, LGTM!

**Comment by Benglin** [comment]
Thanks a lot, @aparajit-pratap!

File: src/DynamoCore/Models/RecordableCommands.cs (hunk 5 above)
Comment (Benglin) [inline src/DynamoCore/Models/RecordableCommands.cs]: ReSharper to the rescue!

**Pull request merged**